### PR TITLE
Add delegated identity API support to spire-api package

### PR DESCRIPTION
**Pull request by EItanya**

### CODE REVIEW

File: spire-api/src/selectors.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            Selector::Generic((k, v)) => SpiffeSelector {
+                r#type: k,
+                value: v,
+            },
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Selector represents a SPIFFE ID selector.
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:29]:
```suggestion
/// Represents various types of SPIFFE identity selectors.
```

File: spire-api/src/selectors.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+pub enum Selector {
+    /// K8s represents a SPIFFE ID selector.
+    K8s(K8s),
+    /// Selector represents a SPIFFE ID selector.
+    Unix(Unix),
+    /// Selector represents a SPIFFE ID selector.
+    Generic((String, String)),
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:36]:
```suggestion
    /// Represents a SPIFFE identity selector based on Kubernetes constructs.
    K8s(K8s),
    /// Represents a SPIFFE identity selector based on Unix system constructs such as PID, GID, and UID.
    Unix(Unix),
    /// Represents a generic SPIFFE identity selector defined by a key-value pair.
    Generic((String, String)),
```

File: spire-api/src/selectors.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+const K8S_SA_TYPE: &str = "sa";
+const K8S_NS_TYPE: &str = "ns";
+
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:41]:
```suggestion

/// Converts Kubernetes selectors to their string representation.
```

File: spire-api/src/selectors.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+#[derive(Debug, Clone)]
+/// K8s is a helper type to create a SPIFFE ID unix process constructs.
+pub enum Unix {
+    /// PID represents the SPIFFE ID selector for a process ID.
+    Pid(u16),
+    /// GID represents the SPIFFE ID selector for a group ID.
+    Gid(u16),
+    /// UID represents the SPIFFE ID selector for a User ID.
+    Uid(u16),
+}
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:83]:
```suggestion
/// Represents SPIFFE identity selectors based on Unix process-related attributes.
#[derive(Debug, Clone)]
pub enum Unix {
    /// Specifies a selector for a Unix process ID (PID).
    Pid(u16),
    /// Specifies a selector for a Unix group ID (GID).
    Gid(u16),
    /// Specifies a selector for a Unix user ID (UID).
    Uid(u16),
}
```

File: spire-api/src/selectors.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    /// ServiceAccount represents the SPIFFE ID selector for a Kubernetes service account.
+    ServiceAccount(String),
+    /// Namespace represents the SPIFFE ID selector for a Kubernetes namespace.
+    Namespace(String),
+}
+
+const UNIX_PID_TYPE: &str = "pid";
+const UNIX_GID_TYPE: &str = "gid";
+const UNIX_UID_TYPE: &str = "uid";
+
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:63]:
```suggestion

/// Converts a Unix selector into a formatted string representation.
```

File: spire-api/src/selectors.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+#[derive(Debug, Clone)]
+/// K8s is a helper type to create a SPIFFE ID selector for Kubernetes.
+pub enum K8s {
+    /// ServiceAccount represents the SPIFFE ID selector for a Kubernetes service account.
+    ServiceAccount(String),
+    /// Namespace represents the SPIFFE ID selector for a Kubernetes namespace.
+    Namespace(String),
+}
```
Comment (maxlambrecht) [inline spire-api/src/selectors.rs:58]:
```suggestion
/// Represents a SPIFFE identity selector for Kubernetes.
pub enum K8s {
    /// SPIFFE identity selector for a Kubernetes service account.
    ServiceAccount(String),
    /// SPIFFE identity selector for a Kubernetes namespace.
    Namespace(String),
}
```

File: spire-api/tests/delegated_identity_api_client_test.rs [hunk 7]
```diff
@@ -1,10 +1,61 @@
...
+            .fetch_x509_bundles()
+            .await
+            .expect("Failed to fetch trust bundles");
+        response
+            .get_bundle(
+                &spiffe::spiffe_id::TrustDomain::new("example.org".as_ref())
+                    .expect("Failed to parse trust domain ="),
+            )
+            .expect("Failed to get bundle");
     }
```
Comment (maxlambrecht) [inline spire-api/tests/delegated_identity_api_client_test.rs:60]: There should be tests exercising the streaming methods.

**Review by maxlambrecht — CHANGES_REQUESTED** [review]
Thanks for this work, @EItanya!

File: spire-api/Cargo.toml [hunk 8]
```diff
@@ -15,10 +15,25 @@ categories = ["cryptography"]
...
-spiffe = { version = "0.3.1", path = "../spiffe" }
+bytes = { version = "1", features = ["serde"] }
+spiffe = { path = "../spiffe" }
 tonic = { version = "0.9", default-features = false, features = ["prost", "codegen", "transport"]}
 prost = { version = "0.11"}
 prost-types = {version = "0.11"}
+tokio = { "version" = "1", features = ["net", "test-util"]}
+tokio-stream = "0.1"
+tower = { version = "0.4", features = ["util"] }
+thiserror = "1.0"
```
[the remark this answers is not in the document]
Comment (EItanya) [reply spire-api/Cargo.toml:26]: Good point, these were copy-pasta, I will remove

File: spire-api/src/agent/delegated_identity.rs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+        let stream = response.into_inner().map(|message| {
+            message
+                .map_err(ClientError::from)
+                .and_then(DelegatedIdentityClient::parse_x509_bundle_set_from_grpc_response)
+        });
+
+        Ok(stream)
+    }
+}
```
[the remark this answers is not in the document]
Comment (EItanya) [reply spire-api/src/agent/delegated_identity.rs:245]: Yes, I didn't need them immediately so I skipped, but happy to add them as well!

File: spire-api/src/agent/delegated_identity.rs [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    /// The function can return an error variant of [`ClientError`] in the following scenarios:
+    ///
+    /// * There's an issue connecting to the Workload API.
+    /// * An error occurs while setting up the stream.
+    ///
+    /// Individual stream items might also be errors if there's an issue processing the response for a specific update.
+    pub async fn stream_x509_svids(
+        mut self,
+        selectors: Vec<Selector>,
+    ) -> Result<impl Stream<Item = Result<X509Svid, ClientError>>, ClientError> {
```
[the remark this answers is not in the document]
Comment (EItanya) [reply spire-api/src/agent/delegated_identity.rs:177]: 100%, this one is a mouthful

[the remark this answers is not in the document]
Comment (EItanya) [reply spire-api/src/agent/delegated_identity.rs:177]: Turns out this isn't supported? https://github.com/rust-lang/rust/issues/89122

[the remark this answers is not in the document]
Comment (maxlambrecht) [reply spire-api/src/agent/delegated_identity.rs:182]: I see, that's still experimental.

**Pull request merged**